src/ImageSharp/Formats/Bmp/BmpDecoderCore.cs

```diff
@@ -209,6 +233,114 @@ private static int CalculatePadding(int width, int componentCount)
         }
 
         /// <summary>
+        /// Looks up color values and builds the image from de-compressed RLE8 data.
+        /// Compresssed RLE8 stream is uncompressed by <see cref="UncompressRle8(int, Buffer{byte})"/>
+        /// </summary>
+        /// <typeparam name="TPixel">The pixel format.</typeparam>
+        /// <param name="pixels">The <see cref="PixelAccessor{TPixel}"/> to assign the palette to.</param>
+        /// <param name="colors">The <see cref="T:byte[]"/> containing the colors.</param>
+        /// <param name="width">The width of the bitmap.</param>
+        /// <param name="height">The height of the bitmap.</param>
+        /// <param name="inverted">Whether the bitmap is inverted.</param>
+        private void ReadRle8<TPixel>(PixelAccessor<TPixel> pixels, byte[] colors, int width, int height, bool inverted)
+            where TPixel : struct, IPixel<TPixel>
+        {
+            var color = default(TPixel);
+            var rgba = default(Rgba32);
+
+            using (var buffer = Buffer<byte>.CreateClean(width * height))
+            {
+                this.UncompressRle8(width, buffer);
+
+                for (int y = 0; y < height; y++)
+                {
+                    int newY = Invert(y, height, inverted);
+                    Span<TPixel> pixelRow = pixels.GetRowSpan(newY);
+                    for (int x = 0; x < width; x++)
+                    {
+                        rgba.Bgr = Unsafe.As<byte, Bgr24>(ref colors[buffer[(y * width) + x] * 4]);
```

Comment:
By using `Buffer2D<byte>` instead of `Buffer<byte>` you can easily avoid the repeating the `y * width` multiplications for each pixel: 
- Use `Span<byte> row = buffer.GetRowSpan(y)` right before the inner loop
- You can index int you row Span by then: `row[x]`

Reply:
I don't think we can do that actually since we do not know which row to write to in `UncompressRle8`

If merged as-is for now.

Reply:
I was going to write to buffer as before (it's still a Buffer<T>), and as Anton says, read whole lines.

Are the CreateClean static methods on Buffer2D<T> missing a 'new' modifier, I couldn't create a clean instance of Buffer2D

Reply:
@georgexcollins The [`CreateClean`](https://github.com/SixLabors/ImageSharp/blob/09729940effcd25c385bd62b24f0b64b144a1fa2/src/ImageSharp/Memory/Buffer2D%7BT%7D.cs#L77) methods should work fine to create an instance that is clean for the required number of elements. 

It is possible however that the underlying array could have populated param type elements at the tail but they are outwith the maximum possible position in the 2D Buffer.

Feel free to submit another PR if you think you can improve on what you have. 😄

Reply:
@JimBobSquarePants I'm talking about reading the uncompressed `buffer` inside the for loops. The inner loop is always using the `y`-th row of `buffer` -> it can read a row span instead.

`UncompressRle8` can process the data into `buffer.Span` as if it was an array.

Reply:
Fool of a Took! @antonfirsov You're absolutely right of course, I can't believe I missed that! 

@georgexcollins I'm gonna add that optimisation as part of #368 to save you the effort.